### ISSUE TITLE
Vectorized zero crossings fail with new frontend

### ISSUE THREAD

**Issue by rfranke** [OpenModelica/OpenModelica#13328]
### Description

The new frontend preserves the reductions in the below example. The code generation does not seem to support this though, omitting some zero crossings and leading to a compilation error.

```Modelica
model Iterator
  constant Integer n = 3;
  input Real[n] u1, u2, u3;
  output Real y1, y2;
  Real[n] x;
equation
  y1 = max({if u1[j] > 1 then u2[j] else u3[j] for j in 1:n});
  y2 = max({if x[j] > 1 then u2[j] else u3[j] for j in 1:n});
  der(x) = u1;
end Iterator;
```

### Steps to Reproduce

Simulate the example. This gives the error:
```
Iterator_05evt.c: In function 'Iterator_function_ZeroCrossings':
Iterator_05evt.c:65:112: error: 'j' undeclared (first use in this function); did you mean 'jn'?
   65 |   tmp0 = GreaterZC((&data->localData[0]->realVars[data->simulationInfo->realVarsIndex[0]] /* x[1] STATE(1) */)[j - 1], 1.0, tmp1, tmp2, data->simulationInfo->storedRelations[0]);
      |                                                                                                                ^
      |                                                                                                                jn
Iterator_05evt.c:65:112: note: each undeclared identifier is reported only once for each function it appears in
Iterator_05evt.c: In function 'Iterator_function_updateRelations':
Iterator_05evt.c:103:114: error: 'j' undeclared (first use in this function); did you mean 'jn'?
  103 |     tmp6 = GreaterZC((&data->localData[0]->realVars[data->simulationInfo->realVarsIndex[0]] /* x[1] STATE(1) */)[j - 1], 1.0, tmp7, tmp8, data->simulationInfo->storedRelations[0]);
      |                                                                                                                  ^
      |                                                                                                                  jn
Iterator_05evt.c: At top level:
cc1: note: unrecognized command-line option '-Wno-parentheses-equality' may have been intended to silence earlier diagnostics
make: *** [<builtin>: Iterator_05evt.o] Error 1
make: *** Waiting for unfinished jobs....
Compilation process failed. Exited with code 2.
```

### Expected Behavior

The translation should generate `n=3` zero crossings for `x`. Instead it creates one zero crossing for x and another one for y, using the undeclared iteration variable `j`.

### Version and OS

 - OpenModelica Version: v1.24 and v1.25.0-dev

**Comment by casella** [comment]
Maybe it should be unrolled during the last scalarization step, unless --newBackend is set.

@perost what do you think?

**Comment by perost** [comment]
The NF simplifies `max({... for ...})` into `max(... for ...)`, so we get:
```Modelica
  y1 = max(if u1[j] > 1.0 then u2[j] else u3[j] for j in 1:3);
  y2 = max(if x[j] > 1.0 then u2[j] else u3[j] for j in 1:3);
```
One possible solution would be to not do this simplification, by replacing:
https://github.com/OpenModelica/OpenModelica/blob/92d56a9bc2d5e55010a4053d55078b4542096a9e/OMCompiler/Compiler/NFFrontEnd/NFSimplifyExp.mo#L334
with
```Modelica
    exp := Expression.CALL(call);
```
But that would just be a workaround for the backend issue, you can still trigger the issue with both the NF and the OF by using e.g. a `max` reduction call. We could of course also unroll it in the NF, but we usually try to avoid that for expressions like this.

So fixing the issue in the backend would be preferable, but I'm not sure how easy that is. The issue seems to be that we get some zero crossing functions where the expressions contain the iterator but without actually being in a for-loop.

**Comment by kabdelhak** [comment]
I fixed the backend part for the new backend in  #13529 and have code to fix it for old backend ready too. unfortunately there are still runtime issues for the new backend and the old backend does not seem to pass the information correctly to simcode. But both should hopefully not be that hard to fix.

Since i need to fix this for the new backend anyway and the old backend port seems straight forward (because i updated sim code structures for new backend already) it should be doable without too much effort.

**Comment by rfranke** [comment]
Thank you @kabdelhak! 
There is some optional treatment of for loops in `BackendDAE.ZERO_CROSSING`
https://github.com/OpenModelica/OpenModelica/blob/28b9a76fbe77155bcdd07629b8dcb77f14ec9ca6/OMCompiler/Compiler/BackEnd/BackendDAE.mo#L659C1-L667C1
I hardly see it used anywhere. Would take over respective extensions in the Cpp code generation.

**Comment by kabdelhak** [comment]
The original pull request that introduced these iterators was #9762, looking at the template changes in `CodegenC.tpl` should suffice for what you want to do. i now have a new pull request that fixes some indexing issues #13535, but as i mentioned you probably wont have those in cpp anyway.

There are remaining issues with the top level inputs for this MWE, but if the inputs are just variables the model works in the new backend with PR #13535. I will try and fix the top level input stuff and then go back to the old backend and see why the iterators don reach SimCode.

@rfranke you might still have the same problem that for some reason the iterators i collect don't reach simcode/template.

**Comment by kabdelhak** [comment]
I fixed the model for the new backend in #13541.

Will now fix it for C-runtime in the old backend (because i understand that one more than cpp) then i will try for cpp. If i don't get far i will ask you @rfranke

**Comment by kabdelhak** [comment]
It runs now c-code old backend with #13544 which should also fix stuff for cpp. the information is now properly available, it just needs to be parsed by the CPP code generation.

unfortunately there are still memory issues, which seem to only occur when freeing after finishing the simulation. A proper correct result file is available. I will still try to fix these memory issues and the cpp code gen.

**Comment by kabdelhak** [comment]
Quick updates today: I finalized the PR. The number of zero crossings was incorrect since it only counted the length of the list not considering single elements might be larger. That is fixed now, no memory issues with OB C-Code anymore.

**Comment by kabdelhak** [comment]
This was the PR that fixed the indexing issue: #13535 we do not use the index saved to the `SIM_VAR` anymore and rather generate it on the fly. not the cleanest solution but it does the job currently

**Comment by casella** [comment]
@rfranke do you need this fix in 1.25.0, or you don't care?
